src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests/System/Runtime/InteropServices/CollectionsMarshalTests.cs

```diff
@@ -505,5 +505,52 @@ private class IntAsObject
             public int Value;
             public int Property { get; set; }
         }
+
+        [Fact]
+        public void ListSetCount()
+        {
+            List<int> list = null!;
+            Assert.Throws<NullReferenceException>(() => CollectionsMarshal.SetCount(list, 3));
+
+            Assert.Throws<ArgumentOutOfRangeException>(() => CollectionsMarshal.SetCount(list, -1));
+
+            list = new();
+            Assert.Throws<ArgumentOutOfRangeException>(() => CollectionsMarshal.SetCount(list, -1));
+
+            CollectionsMarshal.SetCount(list, 5);
+            Assert.Equal(5, list.Count);
+
+            list = new() { 1, 2, 3, 4, 5 };
+            // make sure that size decrease preserves content
+            CollectionsMarshal.SetCount(list, 3);
+            Assert.Equal(3, list.Count);
+            Assert.Throws<ArgumentOutOfRangeException>(() => list[3]);
+            SequenceEquals<int>(CollectionsMarshal.AsSpan(list), new int[] { 1, 2, 3 });
+
+            // make sure that size increase preserves content
+            CollectionsMarshal.SetCount(list, 5);
+            SequenceEquals<int>(CollectionsMarshal.AsSpan(list)[..3], new int[] { 1, 2, 3 });
+
+            // make sure that reallocations preserve content
+            int newCount = list.Capacity * 2;
+            CollectionsMarshal.SetCount(list, newCount);
+            Assert.Equal(newCount, list.Count);
+            SequenceEquals<int>(CollectionsMarshal.AsSpan(list)[..3], new int[] { 1, 2, 3 });
+
+            List<string> listReference = new() { "a", "b", "c", "d", "e" };
+            CollectionsMarshal.SetCount(listReference, 3);
+            // verify that reference types aren't cleared
+            SequenceEquals<string>(CollectionsMarshal.AsSpan(listReference), new string[] { "a", "b", "c" });
```

Comment:
Shouldn't we also test that `"d"` and `"e"` have been cleared in the larger buffer?

Reply:
I did add such test at first but @tannnergooding said that I don't need to test such implementation details since the API doesn't define the content after such downsize and upsize. Should I add it back?

Reply:
Even if we don't document this I wouldn't consider it implementation detail. It's important that the buffer doesn't hold on to invisible references (which won't get removed even if you call the `Clear()` method) so a test validating that we don't regress it would be valuable.

Reply:
@eiriktsarpalis, The general issue is that observing that semantic requires something like `InternalsVisibleTo` or relying on implementation details of how `SetCount` works.

There isn't a way to just observe the backing array of the list from the test, so at best we'd be validating that `SetCount(3)`, followed by `SetCount(5)` "preserves" the value for value types and clears the value for reference types.

We do not similarly test that references get cleared for cases like `Remove`, which are much more important and mainline scenarios.

Reply:
Added tests that verify the underlying buffer is the same.